java/arcs/android/storage/service/BaseStorageChannel.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.service
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withTimeout
+
+/** Base functionality common to all storage channels. */
+abstract class BaseStorageChannel(
+  private val scope: CoroutineScope,
+  private val statisticsSink: BindingContextStatisticsSink
+) : IStorageChannel.Stub() {
+  val actionLauncher = SequencedActionLauncher(scope)
+
+  var listenerToken: Int? = null
+
+  override fun idle(timeoutMillis: Long, resultCallback: IResultCallback?) {
+    // Don't use the SequencedActionLauncher, since we don't want an idle call to wait for other
+    // idle calls to complete.
+    scope.launch {
+      statisticsSink.traceAndMeasure("idle") {
+        resultCallback?.wrapException("idle failed") {
+          checkChannelIsOpen()
+          withTimeout(timeoutMillis) {
+            actionLauncher.waitUntilDone()
+            idleStore()
+          }
+        }
+      }
+    }
+  }
+
+  override fun close(resultCallback: IResultCallback?) {
+    actionLauncher.launch {
+      statisticsSink.traceAndMeasure("MuxedStorageChannel.close") {
+        resultCallback?.wrapException("close failed") {
+          val token = checkNotNull(listenerToken) { "Channel has already been closed" }
+          unregisterFromStore(token)
+          listenerToken = null
+        }
+      }
+    }
+  }
+
+  fun checkChannelIsOpen() {
+    checkNotNull(listenerToken) { "Channel is closed" }
+  }
+
+  abstract suspend fun idleStore()
+
+  abstract suspend fun unregisterFromStore(token: Int)
```

Comment:
I made these two abstract methods in order for `idle` and `close` implementations to be shared by `MuxedStorageChannelImpl` and `StorageChannelImpl`.